File: frontend/src/elements/gui/widgets/table/tableElement.ts
import type { GuiWidgetElementDefinition } from '../../../types';
import TableEditor from './TableEditor';
import TableWidget from '../../../../components/gui/widgets/TableWidget';
import { codeExtension } from '../../../shared/authoredFileName';

/** Display-only, like plot_window: takes rows to show, emits nothing. */
export const tableElement: GuiWidgetElementDefinition = {
  widgetKind: 'table',
  ports: (widget) => {
    const inId = `${widget.id}_in`;
    const label = widget.label || widget.id;
    return {
      inputs: [{ id: inId, name: label, kind: 'input', data_type: 'any', multi: true, required: false, description: '' }],
      outputs: [],
    };
  },
  authoredFile: (widget) => ({ extension: codeExtension(widget), what: 'this transform' }),
  generation: {
    promptField: 'code_prompt',
    targetField: 'code',
    guard: 'Please describe how to turn the incoming data into rows first.',
    success: '✅ Transform generated!',
    promptLabel: 'Prompt',
    promptPlaceholder: 'Describe the rows you want, e.g. one row per file with name, size and date.',
    bodyLabel: 'Optional transform — run(inputs) receives {"value"} and returns {"value"}',
    bodyPlaceholder: 'Leave empty to show the incoming rows as-is.',
    language: true,
    bodyHeight: 90,
  },
  ConfigEditor: TableEditor,
  RuntimeWidget: TableWidget,
};

File: backend/app/elements/gui/widgets/table/table_element.py
"""The `table` GUI widget: display-only, shows rows."""

from __future__ import annotations

from app.elements.base import DisplayWidget, Generation
from app.models.graph import GuiWidgetKind


class TableElement(DisplayWidget):
    """Same contract as the other display widgets: one input port, no output,
    and an optional snippet that reshapes whatever arrives into rows."""

    widget_kind = GuiWidgetKind.TABLE
    config_fields = ("code", "code_prompt")

    def generation(self) -> Generation:
        return Generation(
            kind="code", prompt_field="code_prompt", target_field="code",
            contract=(
                'Must expose run(inputs: dict) -> dict, receiving {"value": <raw incoming data>} '
                'and returning {"value": <table-ready rows>}. Table-ready data is a '
                'JSON-serializable list of objects with the same keys (the keys become the '
                'column headers), or a list of lists whose first row is the header. The app '
                'renders the table itself -- do NOT format it as text and do NOT import '
                'third-party libraries: the code runs in a sandbox with only the standard '
                'library available.'
            ),
            inputs=("value",), outputs=("value",),
            guard="Please describe how to turn the incoming data into rows first.",
            success="✅ Transform generated!",
        )
